otherlibs/fiber/pool.ml

```diff
@@ -20,26 +30,59 @@ let running t k =
   | Open -> k true
   | Closed -> k false
 
-let create () = { mvar = Mvar.create (); status = Open }
+let create () =
+  { tasks = Queue.create (); runner = Awaiting_run; status = Open }
 
 let task t ~f k =
   match t.status with
   | Closed ->
     Code_error.raise "pool is closed. new tasks may not be submitted" []
-  | Open -> Mvar.write t.mvar (Task f) k
-
-let stream t =
-  Stream.In.create (fun () ->
-      let+ next = Mvar.read t.mvar in
-      match next with
-      | Done -> None
-      | Task task -> Some task)
+  | Open -> (
+    Queue.push t.tasks f;
+    match t.runner with
+    | Running | Awaiting_run -> k ()
+    | Awaiting_resume r ->
+      t.runner <- Running;
+      resume r () k)
 
 let stop t k =
   match t.status with
   | Closed -> k ()
-  | Open ->
+  | Open -> (
     t.status <- Closed;
-    Mvar.write t.mvar Done k
+    match t.runner with
+    | Running | Awaiting_run -> k ()
```

Comment:
It looks like we allow to call `stop` before `run` but that seems odd naming-wise. Should `stop` be named `close` instead?

Reply:
(`close` will also match better with the `status`.)

Reply:
Will do that in a subsequent PR.